### PR TITLE
Freeze dependencies

### PR DESCRIPTION
**Pull request by sol**

### CODE REVIEW

**Comment by sol** [comment]
@soenkehahn Can you get your input on this?  Main advantages over freezing via `cabal.config` are:

- we can freeze without doing a `cabal configure`
- we ignore the freeze file if there is no install plan
- in a next step we can freeze git revisions too

File: src/Tinc/Freeze.hs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+sortByName :: [Dependency] -> [Dependency]
+sortByName = sortBy $ comparing f
+  where
+    f :: Dependency -> (String, String)
+    f (Dependency n _) = (map toLower n, n)
+
+toDependency :: Package -> Dependency
+toDependency (Package n (Version v _)) = Dependency {name =  n, version = v}
+
+toConstrain :: Dependency -> Constrain
```
Comment (soenkehahn) [inline src/Tinc/Freeze.hs]: Missing `t` in the name?

File: src/Tinc/Freeze.hs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+data Dependency = Dependency {
+  name :: String
+, version :: String
+} deriving (Eq, Show, Generic)
+
+instance FromJSON Dependency
+
+instance ToJSON Dependency where
+  toJSON = genericToJSON defaultOptions
```
Comment (soenkehahn) [inline src/Tinc/Freeze.hs]: Isn't this the default implementation?

File: src/Tinc/Freeze.hs (hunk 1 above)
Comment (sol) [inline src/Tinc/Freeze.hs]: Good catch.  Thx.

File: src/Tinc/Freeze.hs (hunk 2 above)
Comment (sol) [inline src/Tinc/Freeze.hs]: With `aeson-0.10.0.0` you get a warning if you don't specify it explicitly... :(:(:(

**Comment by sol** [comment]
@soenkehahn Addressed all your comments.  What do you think about the general approach?  What about the name of the freeze file, is `.tinc.freeze` good?  Should it be a dot file or not?  Bundler uses `Gemfile.lock`.

File: src/Tinc/Freeze.hs (hunk 2 above)
Comment (soenkehahn) [inline src/Tinc/Freeze.hs]: Ah, yes, I saw that issue. :(

**Comment by soenkehahn** [comment]
I'm not sure. I think it would be great if it were possible to use `tinc` on normal cabal projects, possibly collaborating with people that use plain `cabal`. So `tinc` should ideally at least consider cabal freeze files when building sandboxes. It would also be desirable to be able to create them, but that can be easily achieved with `tinc && cabal freeze`. Git revisions could also be specified in a `tinc.yaml` file in addition to a cabal freeze file.

But as I said, I'm not sure.

I think I would prefer it not being a dot file. But I have no strong arguments either way.

**Comment by sol** [comment]
1. `tinc` still takes `cabal.config` into account.  I would not want to change that for now.  That means that `cabal freeze` is taken into account.

2. To create a `cabal.config` file, strictly speaking you have to run `tinc && cabal configure && cabal freeze`

3. Yes, I think I agree that it should not be a dot file.  What about `tinc.freeze`?

4. I'm puzzled whether we should skip creating a `tinc.freeze` if there is (a) a `cabal.config` and (b) no `tinc.freeze`?

**Comment by sol** [comment]
Again, regarding (2.), the requirement to do a `cabal configure` is my main reason that I did not freeze dependencies with `cabal`.  I'm not that excited about always running a `cabal config` after creating the sandbox for a couple of reasons (ask, if I should elaborate).

**Pull request merged**